core/tests/test_apps.bitcoin.authorization.py

```diff
@@ -0,0 +1,109 @@
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.com" + int.to_bytes(1, _ROUND_ID_LEN, "big"),
+        )
+
+        self.assertFalse(self.authorization.check_get_ownership_proof(msg))
+
+    def test_ownership_proof_account_path_mismatch(self):
+        # Account path mismatch.
+        msg = GetOwnershipProof(
+            address_n=[H_(49), H_(0), H_(0), 1, 2],
+            coin_name=self.coin.coin_name,
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.com" + int.to_bytes(1, _ROUND_ID_LEN, "big"),
+        )
+
+        self.assertFalse(self.authorization.check_get_ownership_proof(msg))
+
+    def test_ownership_proof_coordinator_mismatch(self):
+        # Coordinator name mismatch.
+        msg = GetOwnershipProof(
+            address_n=[H_(84), H_(0), H_(0), 1, 2],
+            coin_name=self.coin.coin_name,
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.org" + int.to_bytes(1, _ROUND_ID_LEN, "big"),
+        )
+
+        self.assertFalse(self.authorization.check_get_ownership_proof(msg))
+
+    def test_ownership_proof_wrong_round_id(self):
+        # Wrong round ID length.
+        msg = GetOwnershipProof(
+            address_n=[H_(84), H_(0), H_(0), 1, 2],
+            coin_name=self.coin.coin_name,
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.com" + int.to_bytes(1, _ROUND_ID_LEN - 1, "big"),
+        )
+
+        self.assertFalse(self.authorization.check_get_ownership_proof(msg))
+
+        msg = GetOwnershipProof(
+            address_n=[H_(84), H_(0), H_(0), 1, 2],
+            coin_name=self.coin.coin_name,
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.com" + int.to_bytes(1, _ROUND_ID_LEN + 1, "big"),
+        )
+
+        self.assertFalse(self.authorization.check_get_ownership_proof(msg))
+
+    def test_authorize_ownership_proof(self):
+
+        msg = GetOwnershipProof(
+            address_n=[H_(84), H_(0), H_(0), 1, 2],
+            coin_name=self.coin.coin_name,
+            script_type=InputScriptType.SPENDWITNESS,
+            user_confirmation=True,
+            commitment_data=b"www.example.com" + int.to_bytes(1, _ROUND_ID_LEN, "big"),
+        )
+
+        self.assertTrue(self.authorization.check_get_ownership_proof(msg))
+
+    def test_authorize_sign_tx(self):
+
+        msg = SignTx(outputs_count=10, inputs_count=21, coin_name=self.coin.coin_name, lock_time=0)
+
+        self.assertTrue(self.authorization.authorize_sign_tx(msg, 10000))
+        self.assertTrue(self.authorization.authorize_sign_tx(msg, 20000))
+        self.assertFalse(self.authorization.authorize_sign_tx(msg, 10001))
```

Comment:
perhaps add one more `assertTrue(...(10000))`? or isn't that a desired behavior?

Reply:
The second `assertTrue` ensures that you can do repeated approvals using one authorization object, so I don't see the need for a third `assertTrue`. The `assertFalse` ensures that if there is a remaining authorized amount but you exceed it, then the approval will not be issued.

BTW I think I should rename `authorize_sign_tx` -> `approve_sign_tx` to be more consistent with the terminology. These are pretty much synonyms but I try to distinguish "authorize" when referring to some general intention which always comes directly from the user and "approve" when referring to a concrete action that is in line with the intention.

Reply:
> The second `assertTrue` ensures that you can do repeated approvals using one authorization object, so I don't see the need for a third `assertTrue`. The `assertFalse` ensures that if there is a remaining authorized amount but you exceed it, then the approval will not be issued.

a subsequent `assertTrue` would validate that a failed approval did not modify the remaining allowable fee.

> BTW I think I should rename `authorize_sign_tx` -> `approve_sign_tx` to be more consistent with the terminology. These are pretty much synonyms but I try to distinguish "authorize" when referring to some general intention which always comes directly from the user and "approve" when referring to a concrete action that is in line with the intention.

OK by me

Reply:
> a subsequent `assertTrue` would validate that a failed approval did not modify the remaining allowable fee.

Oh, I see. Good point.

Reply:
https://github.com/trezor/trezor-firmware/pull/1127/commits/e9831759a8fc2912e47757487f7befb89f5d272e